src/tools/wasm-ctor-eval.cpp

```diff
@@ -572,13 +572,28 @@ struct CtorEvalExternalInterface : EvallingModuleRunner::ExternalInterface {
   // If |possibleDefiningGlobal| is provided, it is the name of a global that we
   // are in the init expression of, and which can be reused as defining global,
   // if the other conditions are suitable.
-  Expression* getSerialization(const Literal& value,
+  Expression* getSerialization(Literal value,
                                Name possibleDefiningGlobal = Name()) {
     Builder builder(*wasm);
 
-    if (!value.isData()) {
-      // This can be handled normally.
-      return builder.makeConstantExpression(value);
+    // If this is externalized then we want to inspect the inner data, handle
+    // that, and emit a ref.externalize around it as needed. To simplify the
+    // logic here, we save the original (possible externalized) value, and then
+    // look at the internals from here on out.
+    Literal original = value;
+    if (value.type.isRef() && value.type.getHeapType() == HeapType::ext) {
+      value = value.internalize();
+    }
+
+    // An externalized i31 has both original.isData and value.isData as being
+    // false, so we need to check both. That is, we can hand off for normal
+    // handling anything that is not (externalized or not) GC data. GC data is
+    // the only thing that requires special handling here by adding globals as
+    // needed.
+    if (!original.isData() && !value.isData()) {
+      // This can be handled normally. We can pass in the original value here as
+      // makeConstantExpression() knows how to handle externalization itself.
+      return builder.makeConstantExpression(original);
```

Comment:
Eventually we will have to support externref values that actually are external, not just just externalized internal references. (The upstream spec tests depend on this, so we will need to implement it if we ever want to be able to run those.) This code doesn't look forward compatible with that because it seems unlikely that `makeConstantExpression` will be able to do anything with a truly external reference. Should we check more explicitly for i31 here?

Reply:
Makes sense. I added some assertions and now we check that external things are either data or i31.